### PR TITLE
PSCred is being used instead of username and password in UtilityModule

### PR DESCRIPTION
**Pull request by Ajay-MS**

### CODE REVIEW

File: Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1 [hunk 1]
```diff
@@ -72,24 +72,29 @@ function GetScriptToRun
     $serverName = EscapeSpecialChars -str $serverName
     $databaseName = EscapeSpecialChars -str $databaseName
 
+    $securePasswordScript = "`$secureAdminPassword = `"$sqlPassword`" | ConvertTo-SecureString  -AsPlainText -Force"
```
Comment (mvvsubbu) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: You don't need piping, you do this way :  ConvertTo-SecureString 'P@ssw0rd' -AsPlainText -Force

File: Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1 [hunk 2]
```diff
@@ -72,24 +72,29 @@ function GetScriptToRun
...
+    $psCredentialCretionScript = "`$psCredential = New-Object System.Management.Automation.PSCredential (`"$sqlUserName`", `$secureAdminPassword)"
```
Comment (mvvsubbu) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: typo in variable name

File: Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1 [hunk 3]
```diff
@@ -72,24 +72,29 @@ function GetScriptToRun
...
-        $sqlDacpacOnTargetMachinesScript = [string]::Format("{0} {1} ( {2} )", $sqlPackageScript,  [Environment]::NewLine,  $invokeMain)
+        $sqlDacpacOnTargetMachinesScript = [string]::Format("{0} {1} {2} {3} ( {4} )", $sqlPackageScript,  [Environment]::NewLine, $initScript,[Environment]::NewLine,  $invokeMain)
         return $sqlDacpacOnTargetMachinesScript
     }
     else
     {
         $sqlQueryScript = Get-Content TaskModuleSqlUtility\SqlQueryOnTargetMachines.ps1 | Out-String
 
-        $invokeExecute = "Execute-SqlQueryDeployment -taskType `"$taskType`" -sqlFile `"$sqlFile`" -inlineSql `"$inlineSql`" -serverName `"$serverName`" -databaseName `"$databaseName`" -authscheme $authscheme -sqlUsername `"$sqlUsername`" -sqlPassword `"$sqlPassword`" -additionalArguments `"$additionalArguments`""
+        $invokeExecute = "Execute-SqlQueryDeployment -taskType `"$taskType`" -sqlFile `"$sqlFile`" -inlineSql `"$inlineSql`" -serverName `"$serverName`" -databaseName `"$databaseName`" -authscheme $authscheme -sqlPSCredential `$psCredential -additionalArguments `"$additionalArguments`""
```
Comment (mvvsubbu) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: call the variable as sqlServerCredentials and update wherever applicable

File: Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1 [hunk 4]
```diff
@@ -72,24 +72,29 @@ function GetScriptToRun
...
 
         Write-Verbose "Executing main function in SqlQueryOnTargetMachines : $invokeExecute"
-        $sqlScriptOnTargetMachines = [string]::Format("{0} {1} ( {2} )", $sqlQueryScript,  [Environment]::NewLine,  $invokeExecute)
+        $sqlScriptOnTargetMachines = [string]::Format("{0} {1} {2} {3} ( {4} )", $sqlQueryScript,  [Environment]::NewLine, $initScript,[Environment]::NewLine,  $invokeExecute)
```
Comment (mvvsubbu) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: space after comma

Comment (mvvsubbu) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: some places two spaces are there

File: TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1 [hunk 5]
```diff
@@ -394,13 +394,15 @@ function Get-SqlPackageCmdArgs
...
-    [string]$sqlUsername,
-    [string]$sqlPassword,
+    [System.Management.Automation.PSCredential]$sqlPSCredential,
     [string]$connectionString,
     [string]$publishProfile,
     [string]$additionalArguments
     )
 
+    $sqlUsername = $sqlPSCredential.GetNetworkCredential().username
+    $sqlPassword = $sqlPSCredential.GetNetworkCredential().password
```
Comment (mvvsubbu) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1]: is this plain password?

File: TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1 [hunk 6]
```diff
@@ -462,16 +464,15 @@ function Execute-DacpacDeployment
...
+     [System.Management.Automation.PSCredential]$sqlPSCredential,
      [string]$connectionString,
      [string]$publishProfile,
      [string]$additionalArguments
     )
 
     Write-Verbose "Entering script SqlPackageOnTargetMachines.ps1"
     $sqlPackage = Get-SqlPackageOnTargetMachine
-    $sqlPackageArguments = Get-SqlPackageCmdArgs -dacpacFile $dacpacFile -targetMethod $targetMethod -serverName $serverName -databaseName $databaseName -authscheme $authscheme -sqlUsername $sqlUsername -sqlPassword $sqlPassword -connectionString $connectionString -publishProfile $publishProfile -additionalArguments $additionalArguments
+    $sqlPackageArguments = Get-SqlPackageCmdArgs -dacpacFile $dacpacFile -targetMethod $targetMethod -serverName $serverName -databaseName $databaseName -authscheme $authscheme -sqlPSCredential $sqlPSCredential -connectionString $connectionString -publishProfile $publishProfile -additionalArguments $additionalArguments
```
Comment (mvvsubbu) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1]: same as above

File: Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1 [hunk 7]
```diff
@@ -72,24 +72,29 @@ function GetScriptToRun
...
 
+    $securePasswordScript = "`$secureAdminPassword = `"$sqlPassword`" | ConvertTo-SecureString  -AsPlainText -Force"
+    $psCredentialCretionScript = "`$psCredential = New-Object System.Management.Automation.PSCredential (`"$sqlUserName`", `$secureAdminPassword)" 
+    $initScript = [string]::Format("{0} {1} {2}", $securePasswordScript, [Environment]::NewLine,  $psCredentialCretionScript)
+    Write-Verbose "Init script $initScript"
+
     if ($taskType -eq "dacpac")
     {
-        $invokeMain = "Execute-DacpacDeployment -dacpacFile `"$dacpacFile`" -targetMethod $targetMethod -serverName `"$serverName`" -databaseName `"$databaseName`" -authscheme $authscheme -sqlUsername `"$sqlUsername`" -sqlPassword `"$sqlPassword`" -connectionString `"$connectionString`" -publishProfile `"$publishProfile`" -additionalArguments `"$additionalArguments`""
+        $invokeMain = "Execute-DacpacDeployment -dacpacFile `"$dacpacFile`" -targetMethod $targetMethod -serverName `"$serverName`" -databaseName `"$databaseName`" -authscheme $authscheme -sqlPSCredential `$psCredential -connectionString `"$connectionString`" -publishProfile `"$publishProfile`" -additionalArguments `"$additionalArguments`""
```
Comment (mvvsubbu) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: when username and password is empty ?

Comment (mvvsubbu) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: Run gulp test and fix tests. If there are no tests for the new code add it

**Review by mvvsubbu — CHANGES_REQUESTED** [review]

File: TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1 (hunk 5 above)
Comment (Ajay-MS) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1]: Yes getting back plain password

File: Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1 [hunk 8]
```diff
@@ -72,24 +72,31 @@ function GetScriptToRun
...
+    {
+        $securePasswordScript = "`$secureAdminPassword = ConvertTo-SecureString `"$sqlPassword`"  -AsPlainText -Force"
+        $psCredentialCreationScript = "`$sqlServerCredentials = New-Object System.Management.Automation.PSCredential (`"$sqlUserName`", `$secureAdminPassword)" 
+        $initScript = [string]::Format("{0} {1} {2}", $securePasswordScript, [Environment]::NewLine,  $psCredentialCreationScript)
+    }
+
     if ($taskType -eq "dacpac")
     {
-        $invokeMain = "Execute-DacpacDeployment -dacpacFile `"$dacpacFile`" -targetMethod $targetMethod -serverName `"$serverName`" -databaseName `"$databaseName`" -authscheme $authscheme -sqlUsername `"$sqlUsername`" -sqlPassword `"$sqlPassword`" -connectionString `"$connectionString`" -publishProfile `"$publishProfile`" -additionalArguments `"$additionalArguments`""
+        $invokeMain = "Execute-DacpacDeployment -dacpacFile `"$dacpacFile`" -targetMethod $targetMethod -serverName `"$serverName`" -databaseName `"$databaseName`" -authscheme $authscheme -sqlPSCredential `$sqlServerCredentials -connectionString `"$connectionString`" -publishProfile `"$publishProfile`" -additionalArguments `"$additionalArguments`""
```
Comment (mvvsubbu) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: sqlPSCredential change this as well and keep them in sync

File: Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1 [hunk 9]
```diff
@@ -72,24 +72,31 @@ function GetScriptToRun
...
-        $sqlDacpacOnTargetMachinesScript = [string]::Format("{0} {1} ( {2} )", $sqlPackageScript,  [Environment]::NewLine,  $invokeMain)
+        $sqlDacpacOnTargetMachinesScript = [string]::Format("{0} {1} {2} {3} ( {4} )", $sqlPackageScript,  [Environment]::NewLine, $initScript, [Environment]::NewLine,  $invokeMain)
         return $sqlDacpacOnTargetMachinesScript
     }
     else
     {
         $sqlQueryScript = Get-Content TaskModuleSqlUtility\SqlQueryOnTargetMachines.ps1 | Out-String
 
-        $invokeExecute = "Execute-SqlQueryDeployment -taskType `"$taskType`" -sqlFile `"$sqlFile`" -inlineSql `"$inlineSql`" -serverName `"$serverName`" -databaseName `"$databaseName`" -authscheme $authscheme -sqlUsername `"$sqlUsername`" -sqlPassword `"$sqlPassword`" -additionalArguments `"$additionalArguments`""
+        $invokeExecute = "Execute-SqlQueryDeployment -taskType `"$taskType`" -sqlFile `"$sqlFile`" -inlineSql `"$inlineSql`" -serverName `"$serverName`" -databaseName `"$databaseName`" -authscheme $authscheme -sqlPSCredential `$sqlServerCredentials -additionalArguments `"$additionalArguments`""
```
Comment (mvvsubbu) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: same as above

File: Extensions/IISWebAppDeploy/Tests/Tasks/SqlDacpacDeploy/DeployToSqlServer.Tests.ps1 [hunk 10]
```diff
@@ -141,23 +141,24 @@ Describe "Tests for testing GetScriptToRun functionality" {
 
         $script = GetScriptToRun -dacpacFile "sample.dacpac" -targetMethod "server" -serverName "localhost" -databaseName "SampleDB" -authscheme "sqlServerAuthentication" -sqlUserName "sampleuser" -sqlPassword "dummypassword" -connectionString "" -publishProfile "" -additionalArguments "" -taskType "dacpac" -inlineSql "" -sqlFile ""
 
+        Write-Verbose $script
         It "should contain script content and invoke expression" {
             ($script.Contains('Dummy Script')) | Should Be $true
-            ($script.Contains('Execute-DacpacDeployment -dacpacFile "sample.dacpac" -targetMethod server -serverName "localhost" -databaseName "SampleDB" -authscheme sqlServerAuthentication -sqlUsername "sampleuser" -sqlPassword "dummypassword" -connectionString "" -publishProfile "" -additionalArguments ""')) | Should Be $true
+            ($script.Contains('Execute-DacpacDeployment -dacpacFile "sample.dacpac" -targetMethod server -serverName "localhost" -databaseName "SampleDB" -authscheme sqlServerAuthentication -sqlPSCredential $sqlServerCredentials -connectionString "" -publishProfile "" -additionalArguments ""')) | Should Be $true
```
Comment (mvvsubbu) [inline Extensions/IISWebAppDeploy/Tests/Tasks/SqlDacpacDeploy/DeployToSqlServer.Tests.ps1]: same as above comment for sqlPSCredential

File: TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1 [hunk 11]
```diff
@@ -421,6 +420,11 @@ function Get-SqlPackageCmdArgs
 
         if($authscheme -eq "sqlServerAuthentication")
         {
+            if($sqlPSCredential)
```
Comment (mvvsubbu) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1]: name this variable as sqlServerCredentials through out

**Review by mvvsubbu — APPROVED** [review]
Looks fine. You can merge after fixing comments

File: TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1 [hunk 12]
```diff
@@ -421,6 +420,11 @@ function Get-SqlPackageCmdArgs
 
         if($authscheme -eq "sqlServerAuthentication")
         {
+            if($sqlServerCredentials)
+            {
+                $sqlUsername = $sqlServerCredentials.GetNetworkCredential().username
+                $sqlPassword = $sqlServerCredentials.GetNetworkCredential().password
+            }
             $sqlPkgCmdArgs = [string]::Format('{0} /TargetUser:"{1}" /TargetPassword:"{2}"', $sqlPkgCmdArgs, $sqlUsername, $sqlPassword)
```
Comment (kmkumaran) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1]: If the above condition is false, you should not be constructing the sqlPkgCmdArgs.

File: TaskModules/powershell/TaskModuleSqlUtility/SqlQueryOnTargetMachines.ps1 [hunk 13]
```diff
@@ -57,6 +56,11 @@ function Execute-SqlQueryDeployment
 
         if($authscheme -eq "sqlServerAuthentication")
         {
+            if($sqlServerCredentials)
+            {
+                $sqlUsername = $sqlServerCredentials.GetNetworkCredential().username
+                $sqlPassword = $sqlServerCredentials.GetNetworkCredential().password
+            }
             $commandToRun += " -Username `"$sqlUsername`" -Password `"$sqlPassword`" "
```
Comment (kmkumaran) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlQueryOnTargetMachines.ps1]: Same as above. If sqlServerCredentials is not present, you should not proceed to constructing commandToRun.

File: Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1 [hunk 14]
```diff
@@ -72,24 +72,31 @@ function GetScriptToRun
...
+        $invokeMain = "Execute-DacpacDeployment -dacpacFile `"$dacpacFile`" -targetMethod $targetMethod -serverName `"$serverName`" -databaseName `"$databaseName`" -authscheme $authscheme -sqlServerCredentials `$sqlServerCredentials -connectionString `"$connectionString`" -publishProfile `"$publishProfile`" -additionalArguments `"$additionalArguments`""
 
         $sqlPackageScript = Get-Content TaskModuleSqlUtility\SqlPackageOnTargetMachines.ps1 | Out-String
 
         Write-Verbose "Executing main function in SqlPackageOnTargetMachines : $invokeMain"
-        $sqlDacpacOnTargetMachinesScript = [string]::Format("{0} {1} ( {2} )", $sqlPackageScript,  [Environment]::NewLine,  $invokeMain)
+        $sqlDacpacOnTargetMachinesScript = [string]::Format("{0} {1} {2} {3} ( {4} )", $sqlPackageScript,  [Environment]::NewLine, $initScript, [Environment]::NewLine,  $invokeMain)
```
Comment (kmkumaran) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: Why do we need to construct a init script first and then adding it to another script? This becomes little complex.  can't we construct all required items here?

Comment (kmkumaran) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: Do you have to worry about special characters while $initScript is expanded?

Comment (Ajay-MS) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: It's because initScript is common for both if and else block.

Comment (Ajay-MS) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: WinRM SqlDacpac working fine with special characters but have to handle in SQL task for machine group

**Pull request merged**